### PR TITLE
Migrate all spec pages

### PR DESCRIPTION
**Pull request by wbamberg**
This PR is the second of a series of PRs to ship the new spec platform, as part of https://github.com/matrix-org/matrix-doc/issues/2906.

## What's in this PR

This PR includes all the pages for the spec (except the changelog, which will be a new page), converted to Hugo Markdown pages, and all the markup fixed.

What's not included in the PR:

* any Hugo shortcodes that render structured data (like the OpenAPI schemas and event schemas).
* the changelog page and associated templates

These will be coming in subsequent PRs.

## Details of the changes

This is a big PR, but I've tried to structure the commits to make it as reviewable as possible. In this section I'll describe the commits.

### [Add empty page placeholders](https://github.com/matrix-org/matrix-doc/pull/2980/commits/ebc6db233b9a626c4daa69ac6507dae7abf022a1)

This adds MD files for all our pages, in the right place and with the right name and front matter.

### [Add page content as raw Pandoc output](https://github.com/matrix-org/matrix-doc/pull/2980/commits/c924b3246fb315299430d5d90e92bf5e452b568e)

This adds the raw output of the Pandoc conversion to Markdown for everything except the client-server modules. This is the only commit that's not reviewable.

### [Fix headings for Appendices](https://github.com/matrix-org/matrix-doc/pull/2980/commits/19836aa9493b58f8801c6e07f2193171e05aa204)

The original source has Appendices in separate files. In the new platform we're maintaining them as a single file, so we have to adjust the headings levels.

### [Remove duplicate titles](https://github.com/matrix-org/matrix-doc/pull/2980/commits/9d4803c8af52aadac784b5a657438935b3f1e4b5)

In Hugo, page titles are automatically taken from the front matter `title` property and then rendered as H1 headings. Since the original sources have their own H1 headings, we then get duplicate headings in the rendered output. This removes the H1 headings from the sources.

### [Remove 'Table of Contents'](https://github.com/matrix-org/matrix-doc/pull/2980/commits/74adbfc1ec43f8fe745d4f488772f2d944eb0199)

The original sources have a "Table of Contents" line. We are using the sidebar for the ToC.

### [Add support for modules](https://github.com/matrix-org/matrix-doc/pull/2980/commits/228c737f56b50b0dfd4e135a8c5412969a53b326)

This adds the client-server modules. This is complicated because we want to maintain them as separate MD files, but combine them into a single page. So there are a few pieces here:
* add the modules, converted using Pandoc, and with front matter including "weight"
* adjust the heading levels in the module files so they fit into the client-server outline
* [add a simple template that adds them to the document](https://github.com/matrix-org/matrix-doc/pull/2980/commits/228c737f56b50b0dfd4e135a8c5412969a53b326#diff-ead6705ba1fb44c75c1f1fb6b7acf42ab09d7ae2dbc882a98e85c699853f83f3), ordered by "weight" (personally I would recommend that the modules appear in alphabetical order by title, but I've chosen the least-change option of preserving the current ordering).
* [update the main client-server doc](https://github.com/matrix-org/matrix-doc/pull/2980/commits/228c737f56b50b0dfd4e135a8c5412969a53b326#diff-c82b624572233237a687eb9baaf76eafd820da0b362057ed0851593343a37d3b) to:
    1. add the preamble talking about modules and feature profiles and
    2. call this template from the main client-server file.

### [Remove 'unstable' warning](https://github.com/matrix-org/matrix-doc/pull/2980/commits/55aed1d296bc9c2e9db558c23f6dd1ffade1abb3)

We're using the ["version_banner" partial](https://github.com/matrix-org/matrix-doc/blob/master/layouts/partials/version-banner.html) for this now.

### [Fix ASCII diagrams](https://github.com/matrix-org/matrix-doc/pull/2980/commits/6c6bd57ebfd6a52ede63b611a52d3a801888da04)

ASCII diagrams sometimes get mangled by Pandoc. This should fix them all. This is easier to review with whitespace changes excluded: https://github.com/matrix-org/matrix-doc/pull/2980/commits/6c6bd57ebfd6a52ede63b611a52d3a801888da04?w=1.

### [Add syntax highlighting](https://github.com/matrix-org/matrix-doc/pull/2980/commits/ab64bda76dcefcff2fcf4e050427409b9d0068af)

Add code fences and language identifiers so we get code highlighting. Again, easier with whitespace changes excluded: https://github.com/matrix-org/matrix-doc/pull/2980/commits/ab64bda76dcefcff2fcf4e050427409b9d0068af?w=1.

Note that there are a couple of examples where the JSON is technically invalid (generally because it uses an ellipse `...` as a value). In these case I've not added syntax highlighting, because it will likely highlight invalid JSON as an error.

### [Support alerts (notes, warnings, rationales)](https://github.com/matrix-org/matrix-doc/pull/2980/commits/338434bfcdcb72c0da9b413bd66989891ea815db)

Added shortcode/partials for notes, warnings, rationales, and updated all the content to call them.

### [Fix internal links](https://github.com/matrix-org/matrix-doc/pull/2980/commits/4e39200cfaef33a6d2d5a6fb75b7895b81eb5728)

For various reasons, including different names for pages and different code for calculating anchors, internal links are broken. This fixes them.

Note that some links in our spec are to rendered structured data like HTTP APIs and event schemas. We're not yet rendering these, so the targets don't exist. In this commit I have not fixed up links to rendered HTTP APIs, and will do that when we add support for rendering the HTTP APIs. Inconsistently, I *have* tended to fix up links to rendered event schemas. :/. But anyway I'll need to go back and double-check all these when we add support for rendering data.

### [Fix heading levels](https://github.com/matrix-org/matrix-doc/pull/2980/commits/02a41edc767f2fee9a4d62fd30beba597197b40a)

In a few places the heading levels are inconsistent with, for example, H4 coming directly under H2. This fixes that.

### [Fix broken Markdown lists](https://github.com/matrix-org/matrix-doc/pull/2980/commits/f0a4f59bb0571cb5491390f34cee163fdb38a08d)

As @p-e-w noted in https://github.com/matrix-org/matrix-doc/pull/2959#issuecomment-762636762, in a few places the conversion of lists seems to have been broken in a particular way. This fixes that.

### [Extend blockquote style](https://github.com/matrix-org/matrix-doc/pull/2980/commits/afa62628e825ac30a75208d5425611b50fa1ad15)

This fixes https://github.com/matrix-org/matrix-doc/issues/2909 by extending the styling for blockquotes so it is applied for all blockquotes in the main document, not just immediate children of the top level. 

### [Use GFM table syntax instead of raw HTML](https://github.com/matrix-org/matrix-doc/pull/2980/commits/52745160f30fd2bca2964ec8dc8c7d820156f03c)

As title. I couldn't find a way to write GFM tables with multiline cell content, which is a pity. Still it's better than raw HTML.

### [Add content for rooms list page](https://github.com/matrix-org/matrix-doc/pull/2980/commits/3864c11c333b7a6069ef4fc6339ad21799ad96ec)

This just adds basic content for the parent page to the room versions pages.

### [Add example that went missing](https://github.com/matrix-org/matrix-doc/pull/2980/commits/965f573c9e84cdd23512c31ff124b0d1ac15f2a6)

One of the examples in the E2E module didn't make it through the Pandoc conversion. This commit adds it back.

This worried me so I dug into it a bit. It looks as if it happens because in the RST this particular example is marked up like `.. code :: json`, while the others are marked up like `.. code:: json`. Our existing toolchain seems to render this properly, but it seems to confuse Pandoc. This is the only instance I can find in the RST sources where examples are marked up like this.

### [Update proposals document](https://github.com/matrix-org/matrix-doc/pull/2980/commits/79036a34cc1119d6133564a109a641ea5100cb71)

This is maybe one of the more controversial changes. The current "Spec proposals" page includes [a series of tables](https://matrix.org/docs/spec/proposals#tables-of-tracked-proposals), generated by some Python, that list MSCs in various states, using the labels attached to them.

I've opted not to re-create this in the new spec platform. I don't really think it adds anything over just using the GitHub UI (for example, ["spec-pr-missing"](https://github.com/matrix-org/matrix-doc/issues?q=label%3Aproposal+label%3Aspec-pr-missing) MSCs). In fact the GitHub view is better, because it's guaranteed to be always up to date. Supporting them in the new platform would add a significant amount of extra complexity.

So instead I've updated the table of "Lifetime states" with links to the corresponding GitHub search.

Note that we've also previously discussed the idea that this page should not live in the spec itself (https://github.com/matrix-org/matrix-doc/issues/2920).

### [Remove {{apis}} template](https://github.com/matrix-org/matrix-doc/pull/2980/commits/47f37d5ab52c4b32fb1fa1a3afd65ce60a355ab6)

I have opted to remove the `{{apis}}` template and just replace it with a list of links. It's only used in one place, and the version information the template included will be redundant anyway.

### [Remove changelog sections](https://github.com/matrix-org/matrix-doc/pull/2980/commits/86152613b1663338819db8334639c6a9bd95e4ad)

Per-spec changelogs will also be redundant now we do global versioning. In a future PR I'll add a single changelog for the whole spec, roughly following the proposal discussed in https://github.com/matrix-org/matrix-doc/issues/2877.

### [Replace sas-emojis template](https://github.com/matrix-org/matrix-doc/pull/2980/commits/183ecfda0314eb7d93412d38c576c407e0ffa3ea)

Replace the template used to generate the table of SAS emojis with a Hugo template.

### [Fix mangling of sub/superscript in e2e module](https://github.com/matrix-org/matrix-doc/pull/2980/commits/fd658f674f604b8d71d3a872344f6aa4c427cee6)

Fix for one of the issues reported by @p-e-w in https://github.com/matrix-org/matrix-doc/pull/2959#issuecomment-762636762:

> There is a really weird bug where mathematical notation has every second character displayed in bold. For an example, see Client-Server section 12.11.2.2, step 8 in the verification process ("private").

### [Formatting fixes for the authentication section](https://github.com/matrix-org/matrix-doc/pull/2980/commits/ea9fced0925a6d19336a1523c33a28a205ad8a48)

This includes a few fixes to the markup for the section on client authentication:
* the section includes a note referenced using a link like `[1]`. This got broken in the Pandoc conversion. Since this is the only place in the spec we were using this form of notes, I replaced with with an inline note box, to be consistent.
* the heading levels were off in this section. This commit fixes that.
* this section used tables for common info about the auth types and client types. They got mangled in the conversion and this commit fixes that.
* the old spec used a `dl` for the contents of the server response in https://matrix.org/docs/spec/client_server/unstable#user-interactive-api-in-the-rest-api. `dl` isn't supported in Markdown, so I replaced this with `li`.

There are a few other places in the spec where we use `dl` - we should probably address them in this PR? But the PR is getting so huge at this point that I'd like review anyway.

## H7/H8

One thing missing from the PR is handling for our `h7` and `h8` problem. There's an issue for this: https://github.com/matrix-org/matrix-doc/issues/2898 in which I listed three options for dealing with it:

* **option 1: work around the existing structure**: we can of course just embed HTML in the Markdown source. So we could have something like `<div class="h7">Default Override Rules</div>` and update things like the ToC code to include it. There will be several different places to implement custom behaviour (ToC, auto-numbering, anchor and anchor-link icon) it may not work well in some edge cases and there will be problems with accessibility (https://css-tricks.com/how-to-section-your-html/#what-happens-when-you-need-h7).

* **option 2: remove the h7/h8**: we make minimal changes to the spec, to remove the h7/h8.

* **option3: restructure the spec**: we restructure the spec so it doesn't require such deep nesting.

I don't think we should go with option 1. I think we should learn to work with only 6 levels of heading. Option 3 sounds like the best outcome but it potentially starts to get into semantic changes rather than just editorial/formatting fixes, so I think it would take a bit more careful discussion, and it would be much easier to discuss that in a separate PR.

For reference, there are all the H7/H8 occurrences:

```
./content/client-server-api/modules/secrets.md:####### `m.pbkdf2`
./content/client-server-api/modules/instant_messaging.md:####### Fallbacks and event representation
./content/client-server-api/modules/instant_messaging.md:######## Stripping the fallback
./content/client-server-api/modules/instant_messaging.md:######## Fallback for `m.text`, `m.notice`, and unrecognised message types
./content/client-server-api/modules/instant_messaging.md:######## Fallback for `m.emote`
./content/client-server-api/modules/instant_messaging.md:######## Fallback for `m.image`, `m.video`, `m.audio`, and `m.file`
./content/client-server-api/modules/push.md:####### Tweaks
./content/client-server-api/modules/push.md:####### Default Override Rules
./content/client-server-api/modules/push.md:######## `.m.rule.master`
./content/client-server-api/modules/push.md:######## `.m.rule.suppress_notices`
./content/client-server-api/modules/push.md:######## `.m.rule.invite_for_me`
./content/client-server-api/modules/push.md:######## `.m.rule.member_event`
./content/client-server-api/modules/push.md:######## `.m.rule.contains_display_name`
./content/client-server-api/modules/push.md:######## `.m.rule.tombstone`
./content/client-server-api/modules/push.md:######## `.m.rule.roomnotif`
./content/client-server-api/modules/push.md:####### Default Content Rules
./content/client-server-api/modules/push.md:######## `.m.rule.contains_user_name`
./content/client-server-api/modules/push.md:####### Default Underride Rules
./content/client-server-api/modules/push.md:######## `.m.rule.call`
./content/client-server-api/modules/push.md:######## `.m.rule.encrypted_room_one_to_one`
./content/client-server-api/modules/push.md:######## `.m.rule.room_one_to_one`
./content/client-server-api/modules/push.md:######## `.m.rule.message`
./content/client-server-api/modules/push.md:######## `.m.rule.encrypted`
```

### CODE REVIEW

**Review by turt2live — COMMENTED** [review]
Looks good overall. I think the only content that disappeared is the section in the proposals intro, but I might also be missing where it went. 

Thank you so much for working on this - all 11.7k lines are amazing improvements :D

File: content/client-server-api/modules/end_to_end_encryption.md [hunk 1]
```diff
@@ -0,0 +1,1584 @@
...
+    verification.
+4.  Alice's client ensures it has a copy of Bob's device key.
+5.  Alice's device sends Bob's device an `m.key.verification.start`
+    message.
+6.  Bob's device receives the message and selects a key agreement
+    protocol, hash algorithm, message authentication code, and SAS
+    method supported by Alice's device.
+7.  Bob's device ensures it has a copy of Alice's device key.
+8.  Bob's device creates an ephemeral Curve25519 key pair
+    (*K*<sub>*B*</sub><sup>*p**r**i**v**a**t**e*</sup>, *K*<sub>*B*</sub><sup>*p**u**b**l**i**c*</sup>),
```
[the remark this answers is not in the document]
Comment (turt2live) [reply content/client-server-api/modules/end_to_end_encryption.md]: answer: no we don't, you fixed this :)

File: content/proposals.md [hunk 2]
```diff
@@ -479,45 +473,9 @@ In summary:
...
     the Java package naming convention. The foundation's preferred
-    vendor prefix is <span class="title-ref">org.matrix</span>.
+    vendor prefix is *org.matrix*.
 -   The vendor prefixes, unstable feature flags, and unstable endpoints
     should be included in the MSC, though the MSC MUST be written in a
     way that proposes new stable endpoints. Typically this is solved by
     a small table at the bottom mapping the various values from stable
     to unstable.
-
-# Proposal Tracking
```
[the remark this answers is not in the document]
Comment (wbamberg) [reply content/proposals.md]: As discussed, can we merge this PR and add these tables back in a subsquent PR, if people miss them?

**Comment by wbamberg** [comment]
Thanks for the review! I've fixed those two formatting issues.

File: content/proposals.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (turt2live) [reply content/proposals.md]: Yup - have recorded https://github.com/matrix-org/matrix-doc/issues/2987 to track it

**Review by turt2live — APPROVED** [review]
Thanks :D

**Pull request merged**